node/core/pvf/src/execute/worker.rs

```diff
@@ -359,7 +366,13 @@ fn validate_using_artifact(
 		//         [`executor_intf::prepare`].
 		executor.execute(artifact_path.as_ref(), params)
 	} {
-		Err(err) => return Response::format_invalid("execute", &err),
+		Err(err) =>
+			return if err.contains("failed to open file: No such file or directory") {
```

Comment:
This really requires a refactor changing the error type to something sensible, like an `enum`.  Matching on a string is way too error prone. Something changes the message, localization, ....

Reply:
Very likely to come from substrate, it's full of string errors. I agree that matching against strings is no-go, but otherwise we'd have to halt the pr.

Reply:
> This really requires a refactor changing the error type to something sensible, like an `enum`

Vote up, I've already raised that concern somewhere... Many errors coming from Substrate are not sensible at all. Also agree that string matching makes no good.

@mrcnski a (probably stupid) idea: until we have an `enum` error from Substrate, would it be better not to rely on its string errors but to check for the file existence ourselves? Should be simple enough. Of course, it introduces a race condition, but still better than parsing strings. Also, nobody guarantees that the file persists between the moments when it is open and when it is read, so that kind of race condition already exists anyway.

Reply:
Yeah, it comes from Substrate. Didn't think about localization. 😬 I considered just treating `RuntimeConstruction` itself as an internal error, but seems it's also used for some case where wasm runs out of memory, which would be a problem with the PVF itself. [link](https://github.com/paritytech/substrate/blob/master/client/executor/src/integration_tests/mod.rs#L824)

Checking for the file existence seems sensible to me...

Reply:
Problems with the PVF itself we agreed are also no reason to raise a dispute, since we have pre-checking enabled. Basically any error that is independent of the candidate at hand should not be cause for a dispute.

Reply:
Let's in any case create a ticket for fixing those string errors - or at least the one in question right now.

Reply:
Looks like `"output exceeds bounds of wasm memory"` should not be an issue with runtime construction, but is actually indicative of a malicious PVF. And I think this would get past pre-checking, since that just compiles and doesn't execute.

https://github.com/paritytech/substrate/blob/master/client/executor/wasmtime/src/runtime.rs#L772-L780

```rs
// Do a length check before allocating. The returned output should not be bigger than the
// available WASM memory. Otherwise, a malicious parachain can trigger a large allocation,
// potentially causing memory exhaustion.
//
// Get the size of the WASM memory in bytes.
let memory_size = ctx.as_context().data().memory().data_size(ctx);
if checked_range(output_ptr as usize, output_len as usize, memory_size).is_none() {
    Err(WasmError::Other("output exceeds bounds of wasm memory".into()))?
}
```

(I used `WasmError::Other` here to match the other errors in the file, without realizing it gets converted to `RuntimeConstruction`. 🤷‍♂️)

Anyway, basically, this one `"output exceeds bounds of wasm memory"` case is deterministic and we should definitely vote against. If we gave it a new separate enum in Substrate, then we could treat the existing `RuntimeConstruction` as an internal error.

Reply:
Raised a small fix for the output-bounds case [here](https://github.com/paritytech/substrate/pull/13841), but I'm still not confident that `RuntimeConstruction` is always a transient error and don't think we should use it. Unless we treat it as a new "possibly transient" case, meaning we retry and dispute if it happens again. 🤷‍♂️

For the file-not-found case, there is not a clear way to fix the error story on the Substrate side. Just having another check here for file existence should be enough.

Reply:
I am fine with the check, please reference the substrate issue though in a comment. This way, readers understand why we did it this way and we can reevaluate once the issue is fixed.